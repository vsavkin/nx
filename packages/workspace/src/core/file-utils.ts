import { execSync } from 'child_process';
import * as fs from 'fs';
import { readFileSync } from 'fs';
import * as path from 'path';
import { extname, join } from 'path';
import { NxArgs } from '../command-line/utils';
import { WorkspaceResults } from '../command-line/workspace-results';
import { appRootPath } from '../utilities/app-root';
import { fileExists, readJsonFile } from '../utilities/fileutils';
import { jsonDiff } from '../utilities/json-diff';
import { ProjectGraphNode } from './project-graph';
import { Environment, NxJson } from './shared-interfaces';
import { defaultFileHasher } from './hasher/file-hasher';
import { performance } from 'perf_hooks';
import { toOldFormatOrNull, Workspaces } from '@nrwl/tao/src/shared/workspace';

const ignore = require('ignore');

export interface FileData {
  file: string;
  hash: string;
  ext: string;
}

export interface Change {
  type: string;
}

export interface FileChange<T extends Change = Change> extends FileData {
  getChanges: () => T[];
}

export class WholeFileChange implements Change {
  type = 'WholeFileChange';
}

export function isWholeFileChange(change: Change): change is WholeFileChange {
  return change.type === 'WholeFileChange';
}

export function calculateFileChanges(
  files: string[],
  nxArgs?: NxArgs,
  readFileAtRevision: (
    f: string,
    r: void | string
  ) => string = defaultReadFileAtRevision,
  ignore = getIgnoredGlobs()
): FileChange[] {
  if (ignore) {
    files = files.filter((f) => !ignore.ignores(f));
  }

  return files.map((f) => {
    const ext = extname(f);
    const hash = defaultFileHasher.hashFile(f);

    return {
      file: f,
      ext,
      hash,
      getChanges: (): Change[] => {
        if (!nxArgs) {
          return [new WholeFileChange()];
        }

        if (nxArgs.files && nxArgs.files.includes(f)) {
          return [new WholeFileChange()];
        }
        switch (ext) {
          case '.json':
            const atBase = readFileAtRevision(f, nxArgs.base);
            const atHead = readFileAtRevision(f, nxArgs.head);

            try {
              return jsonDiff(JSON.parse(atBase), JSON.parse(atHead));
            } catch (e) {
              return [new WholeFileChange()];
            }
          default:
            return [new WholeFileChange()];
        }
      },
    };
  });
}

export const TEN_MEGABYTES = 1024 * 10000;

function defaultReadFileAtRevision(
  file: string,
  revision: void | string
): string {
  try {
    const fileFullPath = `${appRootPath}${path.sep}${file}`;
    const gitRepositoryPath = execSync('git rev-parse --show-toplevel')
      .toString()
      .trim();
    const filePathInGitRepository = path
      .relative(gitRepositoryPath, fileFullPath)
      .split(path.sep)
      .join('/');
    return !revision
      ? readFileSync(file).toString()
      : execSync(`git show ${revision}:${filePathInGitRepository}`, {
          maxBuffer: TEN_MEGABYTES,
        })
          .toString()
          .trim();
  } catch {
    return '';
  }
}

function getFileData(filePath: string): FileData {
  const file = path.relative(appRootPath, filePath).split(path.sep).join('/');
  return {
    file: file,
    hash: defaultFileHasher.hashFile(filePath),
    ext: path.extname(filePath),
  };
}

export function allFilesInDir(
  dirName: string,
  recurse: boolean = true
): FileData[] {
  const ignoredGlobs = getIgnoredGlobs();
  const relDirName = path.relative(appRootPath, dirName);
  if (relDirName && ignoredGlobs.ignores(relDirName)) {
    return [];
  }

  let res = [];
  try {
    fs.readdirSync(dirName).forEach((c) => {
      const child = path.join(dirName, c);
      if (ignoredGlobs.ignores(path.relative(appRootPath, child))) {
        return;
      }
      try {
        const s = fs.statSync(child);
        if (!s.isDirectory()) {
          // add starting with "apps/myapp/..." or "libs/mylib/..."
          res.push(getFileData(child));
        } else if (s.isDirectory() && recurse) {
          res = [...res, ...allFilesInDir(child)];
        }
      } catch (e) {}
    });
  } catch (e) {}
  return res;
}

function getIgnoredGlobs() {
  const ig = ignore();
  ig.add(readFileIfExisting(`${appRootPath}/.gitignore`));
  ig.add(readFileIfExisting(`${appRootPath}/.nxignore`));
  return ig;
}

function readFileIfExisting(path: string) {
  return fs.existsSync(path) ? fs.readFileSync(path, 'UTF-8').toString() : '';
}

export function readWorkspaceJson(): any {
  const ws = new Workspaces(appRootPath);
  return ws.readWorkspaceConfiguration();
}

export function readWorkspaceConfig(opts: { format: 'angularCli' | 'nx' }) {
  const json = readWorkspaceJson();
  if (opts.format === 'angularCli') {
    const formatted = toOldFormatOrNull(json);
    return formatted ? formatted : json;
  } else {
    return json;
  }
}

export function workspaceFileName() {
  if (fileExists(`${appRootPath}/angular.json`)) {
    return 'angular.json';
  } else {
    return 'workspace.json';
  }
}

export type FileRead = (s: string) => string;

export function defaultFileRead(filePath: string): string | null {
  return readFileSync(join(appRootPath, filePath), 'UTF-8');
}

export function readPackageJson(): any {
  return readJsonFile(`${appRootPath}/package.json`);
}

export function readNxJson(): NxJson {
  const config = readJsonFile<NxJson>(`${appRootPath}/nx.json`);
  if (!config.npmScope) {
    throw new Error(`nx.json must define the npmScope property.`);
  }
  return config;
}

export function workspaceLayout(): { appsDir: string; libsDir: string } {
  const nxJson = readNxJson();
  const appsDir =
    (nxJson.workspaceLayout && nxJson.workspaceLayout.appsDir) || 'apps';
  const libsDir =
    (nxJson.workspaceLayout && nxJson.workspaceLayout.libsDir) || 'libs';
  return { appsDir, libsDir };
}

// TODO: Make this list extensible
export function rootWorkspaceFileNames(): string[] {
  return [`package.json`, workspaceFileName(), `nx.json`, `tsconfig.base.json`];
}

export function rootWorkspaceFileData(): FileData[] {
  return rootWorkspaceFileNames().map((f) =>
    getFileData(`${appRootPath}/${f}`)
  );
}

export function readWorkspaceFiles(): FileData[] {
  performance.mark('read workspace files:start');

  if (defaultFileHasher.usesGitForHashing) {
    const ignoredGlobs = getIgnoredGlobs();
    const r = defaultFileHasher.workspaceFiles
      .filter((f) => !ignoredGlobs.ignores(f))
      .map((f) => getFileData(`${appRootPath}/${f}`));
    performance.mark('read workspace files:end');
    performance.measure(
      'read workspace files',
      'read workspace files:start',
      'read workspace files:end'
    );
    r.sort((x, y) => x.file.localeCompare(y.file));
    return r;
  } else {
    const r = [];
    r.push(...rootWorkspaceFileData());

    // Add known workspace files and directories
    r.push(...allFilesInDir(appRootPath, false));
    r.push(...allFilesInDir(`${appRootPath}/tools`));
    const wl = workspaceLayout();
    r.push(...allFilesInDir(`${appRootPath}/${wl.appsDir}`));
    if (wl.appsDir !== wl.libsDir) {
      r.push(...allFilesInDir(`${appRootPath}/${wl.libsDir}`));
    }
    performance.mark('read workspace files:end');
    performance.measure(
      'read workspace files',
      'read workspace files:start',
      'read workspace files:end'
    );
    r.sort((x, y) => x.file.localeCompare(y.file));
    return r;
  }
}

export function readEnvironment(
  target: string,
  projects: Record<string, ProjectGraphNode>
): Environment {
  const nxJson = readNxJson();
  const workspaceJson = readWorkspaceJson();
  const workspaceResults = new WorkspaceResults(target, projects);

  return { nxJson, workspaceJson, workspaceResults };
}

export function normalizedProjectRoot(p: ProjectGraphNode): string {
  if (p.data && p.data.root) {
    const path = p.data.root.split('/').filter((v) => !!v);
    if (path.length === 1) {
      return path[0];
    }
    // Remove the first part of the path, usually 'libs'
    return path.slice(1).join('/');
  } else {
    return '';
  }
}

export function filesChanged(a: FileData[], b: FileData[]) {
  if (a.length !== b.length) return true;

  for (let i = 0; i < a.length; ++i) {
    if (a[i].file !== b[i].file) return true;
    if (a[i].hash !== b[i].hash) return true;
  }
  return false;
}
