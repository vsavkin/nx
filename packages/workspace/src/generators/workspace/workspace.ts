import {
  generateFiles,
  Tree,
  updateJson,
  convertNxGenerator,
  names,
  writeJson,
} from '@nrwl/devkit';
import { Schema } from './schema';
import {
  angularCliVersion,
  eslintVersion,
  nxVersion,
  prettierVersion,
  typescriptVersion,
} from '../../utils/versions';
import { readFileSync } from 'fs';
import { join as pathJoin } from 'path';

export const DEFAULT_NRWL_PRETTIER_CONFIG = {
  singleQuote: true,
};

const decorateAngularClI = (host: Tree) => {
  const decorateCli = readFileSync(
    pathJoin(__dirname as any, '..', 'utils', 'decorate-angular-cli.js__tmpl__')
  ).toString();
  host.write('decorate-angular-cli.js', decorateCli);
};

function setWorkspaceLayoutProperties(tree: Tree, options: Schema) {
  updateJson(tree, 'nx.json', (json) => {
    if (options.layout === 'packages') {
      json.workspaceLayout = {
        appsDir: 'packages',
        libsDir: 'packages',
      };
    }
    return json;
  });
}

function createAppsAndLibsFolders(host: Tree, options: Schema) {
  if (options.layout === 'packages') {
    host.write('packages/.gitkeep', '');
  } else {
    host.write('apps/.gitkeep', '');
    host.write('libs/.gitkeep', '');
  }
}

function createFiles(host: Tree, options: Schema) {
  const npmScope = options.npmScope ? options.npmScope : options.name;
  const formattedNames = names(options.name);
  generateFiles(host, pathJoin(__dirname, './files'), '', {
    formattedNames,
    dot: '.',
    tmpl: '',
    workspaceFile: options.cli === 'angular' ? 'angular' : 'workspace',
    cliCommand: options.cli === 'angular' ? 'ng' : 'nx',
    nxCli: false,
    typescriptVersion,
    prettierVersion,
    eslintVersion,
    // angular cli is used only when workspace schematics is added to angular cli
    angularCliVersion,
    ...(options as object),
    nxVersion,
    npmScope,
  });
}

function createPrettierrc(host: Tree) {
  writeJson(host, '.prettierrc', DEFAULT_NRWL_PRETTIER_CONFIG);
}

export function workspaceGenerator(host: Tree, options: Schema) {
  if (!options.name) {
    throw new Error(`Invalid options, "name" is required.`);
  }
  createFiles(host, options);
  createPrettierrc(host);
  if (options.cli === 'angular') {
    decorateAngularClI(host);
  }
  setWorkspaceLayoutProperties(host, options);
  createAppsAndLibsFolders(host, options);
}

export const workspaceSchematic = convertNxGenerator(workspaceGenerator);
