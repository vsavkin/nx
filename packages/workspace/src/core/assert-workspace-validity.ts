import { workspaceFileName } from './file-utils';
import {
  ImplicitJsonSubsetDependency,
  NxJson,
} from '@nrwl/workspace/src/core/shared-interfaces';
import { output } from '../utilities/output';

export function assertWorkspaceValidity(workspaceJson, nxJson: NxJson) {
  const workspaceJsonProjects = Object.keys(workspaceJson.projects);
  const nxJsonProjects = Object.keys(nxJson.projects);

  if (minus(workspaceJsonProjects, nxJsonProjects).length > 0) {
    output.error({
      title: 'Configuration Error',
      bodyLines: [
        `${workspaceFileName()} and nx.json are out of sync. The following projects are missing in nx.json: ${minus(
          workspaceJsonProjects,
          nxJsonProjects
        ).join(', ')}`,
      ],
    });

    process.exit(1);
  }

  if (minus(nxJsonProjects, workspaceJsonProjects).length > 0) {
    output.error({
      title: 'Configuration Error',
      bodyLines: [
        `${workspaceFileName()} and nx.json are out of sync. The following projects are missing in ${workspaceFileName()}: ${minus(
          nxJsonProjects,
          workspaceJsonProjects
        ).join(', ')}`,
      ],
    });

    process.exit(1);
  }

  const projects = {
    ...workspaceJson.projects,
    ...nxJson.projects,
  };

  const invalidImplicitDependencies = new Map<string, string[]>();

  Object.entries<'*' | string[] | ImplicitJsonSubsetDependency>(
    nxJson.implicitDependencies || {}
  )
    .reduce((acc, entry) => {
      function recur(value, acc = []) {
        if (value === '*') {
          // do nothing since '*' is calculated and always valid.
        } else if (Array.isArray(value)) {
          acc.push([entry[0], value]);
        } else {
          Object.values(value).forEach((v) => {
            recur(v, acc);
          });
        }
      }
      recur(entry[1], acc);
      return acc;
    }, [])
    .reduce((map, [filename, projectNames]: [string, string[]]) => {
      detectAndSetInvalidProjectValues(map, filename, projectNames, projects);
      return map;
    }, invalidImplicitDependencies);

  nxJsonProjects
    .filter((nxJsonProjectName) => {
      const project = nxJson.projects[nxJsonProjectName];
      return !!project.implicitDependencies;
    })
    .reduce((map, nxJsonProjectName) => {
      const project = nxJson.projects[nxJsonProjectName];
      detectAndSetInvalidProjectValues(
        map,
        nxJsonProjectName,
        project.implicitDependencies,
        projects
      );
      return map;
    }, invalidImplicitDependencies);

  if (invalidImplicitDependencies.size === 0) {
    return;
  }

  let message = `The following implicitDependencies specified in nx.json are invalid:
  `;
  invalidImplicitDependencies.forEach((projectNames, key) => {
    const str = `  ${key}
    ${projectNames.map((projectName) => `    ${projectName}`).join('\n')}`;
    message += str;
  });

  output.error({
    title: 'Configuration Error',
    bodyLines: [message],
  });

  process.exit(1);
}

function detectAndSetInvalidProjectValues(
  map: Map<string, string[]>,
  sourceName: string,
  desiredProjectNames: string[],
  validProjects: any
) {
  const invalidProjects = desiredProjectNames.filter(
    (projectName) => !validProjects[projectName]
  );
  if (invalidProjects.length > 0) {
    map.set(sourceName, invalidProjects);
  }
}

function minus(a: string[], b: string[]): string[] {
  const res = [];
  a.forEach((aa) => {
    if (!b.find((bb) => bb === aa)) {
      res.push(aa);
    }
  });
  return res;
}
