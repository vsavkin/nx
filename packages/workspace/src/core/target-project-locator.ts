import { resolveModuleByImport } from '../utilities/typescript';
import { defaultFileRead, FileRead, normalizedProjectRoot } from './file-utils';
import {
  ProjectGraphNode,
  ProjectGraphNodeRecords,
} from './project-graph/project-graph-models';
import {
  getSortedProjectNodes,
  isNpmProject,
  isWorkspaceProject,
} from './project-graph';
import { isRelativePath, parseJsonWithComments } from '../utilities/fileutils';
import { dirname, join } from 'path';
import { appRootPath } from '@nrwl/workspace/src/utilities/app-root';

export class TargetProjectLocator {
  private sortedProjects = getSortedProjectNodes(this.nodes);

  private sortedWorkspaceProjects = this.sortedProjects
    .filter(isWorkspaceProject)
    .map(
      (node) =>
        ({
          ...node,
          data: {
            ...node.data,
            normalizedRoot: normalizedProjectRoot(node),
          },
        } as ProjectGraphNode)
    );
  private npmProjects = this.sortedProjects.filter(isNpmProject);
  private tsConfigPath = this.getRootTsConfigPath();
  private absTsConfigPath = join(appRootPath, this.tsConfigPath);
  private paths = parseJsonWithComments(this.fileRead(this.tsConfigPath))
    ?.compilerOptions?.paths;
  private typescriptResolutionCache = new Map<string, string | null>();
  private npmResolutionCache = new Map<string, string | null>();

  constructor(
    private nodes: ProjectGraphNodeRecords,
    private fileRead: FileRead = defaultFileRead
  ) {}

  /**
   * Find a project based on its import
   *
   * @param importExpr
   * @param filePath
   * @param npmScope
   *  Npm scope shouldn't be used finding a project, but, to improve backward
   *  compatibility, we fallback to checking the scope.
   *  This happens in cases where someone has the dist output in their tsconfigs
   *  and typescript will find the dist before the src.
   */
  findProjectWithImport(
    importExpr: string,
    filePath: string,
    npmScope: string
  ): string {
    const normalizedImportExpr = importExpr.split('#')[0];

    if (isRelativePath(normalizedImportExpr)) {
      const resolvedModule = join(dirname(filePath), normalizedImportExpr);
      return this.findProjectOfResolvedModule(resolvedModule);
    }

    const npmProject = this.findNpmPackage(importExpr);
    if (npmProject) {
      return npmProject;
    }

    if (this.paths && this.paths[normalizedImportExpr]) {
      for (let p of this.paths[normalizedImportExpr]) {
        const maybeResolvedProject = this.findProjectOfResolvedModule(p);
        if (maybeResolvedProject) {
          return maybeResolvedProject;
        }
      }
    }

    const resolvedModule = this.typescriptResolutionCache.has(
      normalizedImportExpr
    )
      ? this.typescriptResolutionCache.get(normalizedImportExpr)
      : resolveModuleByImport(
          normalizedImportExpr,
          filePath,
          this.absTsConfigPath
        );

    this.typescriptResolutionCache.set(normalizedImportExpr, resolvedModule);
    if (resolvedModule) {
      const resolvedProject = this.findProjectOfResolvedModule(resolvedModule);

      if (resolvedProject) {
        return resolvedProject;
      }
    }

    const importedProject = this.sortedWorkspaceProjects.find((p) => {
      const projectImport = `@${npmScope}/${p.data.normalizedRoot}`;
      return normalizedImportExpr.startsWith(projectImport);
    });

    return importedProject?.name;
  }

  private findNpmPackage(npmImport: string) {
    if (this.npmResolutionCache.has(npmImport)) {
      return this.npmResolutionCache.get(npmImport);
    } else {
      const pkgName = this.npmProjects.find(
        (pkg) =>
          npmImport === pkg.data.packageName ||
          npmImport.startsWith(pkg.data.packageName + '/')
      )?.name;
      this.npmResolutionCache.set(npmImport, pkgName);
      return pkgName;
    }
  }

  private findProjectOfResolvedModule(resolvedModule: string) {
    const importedProject = this.sortedWorkspaceProjects.find((p) => {
      return resolvedModule.startsWith(p.data.root);
    });

    return importedProject?.name;
  }

  private getRootTsConfigPath() {
    try {
      this.fileRead('tsconfig.base.json');
      return 'tsconfig.base.json';
    } catch (e) {
      return 'tsconfig.json';
    }
  }
}
