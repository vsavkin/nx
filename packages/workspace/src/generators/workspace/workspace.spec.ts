import { readJson, Tree } from '@nrwl/devkit';
import { workspaceGenerator } from './workspace';
import { createTree } from '@nrwl/devkit/testing';
import { NxJson } from '../../core/shared-interfaces';

describe('@nrwl/workspace:workspace', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTree();
  });

  it('should create files', async () => {
    workspaceGenerator(tree, {
      name: 'proj',
      cli: 'nx',
      layout: 'apps-and-libs',
      defaultBase: 'main',
    });
    expect(tree.exists('/nx.json')).toBe(true);
    expect(tree.exists('/workspace.json')).toBe(true);
    expect(tree.exists('/.prettierrc')).toBe(true);
    expect(tree.exists('/.prettierignore')).toBe(true);
  });

  it('should create nx.json', async () => {
    workspaceGenerator(tree, {
      name: 'proj',
      cli: 'nx',
      layout: 'apps-and-libs',
      defaultBase: 'master',
    });
    const nxJson = readJson<NxJson>(tree, '/nx.json');
    expect(nxJson).toEqual({
      npmScope: 'proj',
      affected: {
        defaultBase: 'master',
      },
      implicitDependencies: {
        'workspace.json': '*',
        'package.json': {
          dependencies: '*',
          devDependencies: '*',
        },
        'tsconfig.base.json': '*',
        'tslint.json': '*',
        '.eslintrc.json': '*',
        'nx.json': '*',
      },
      tasksRunnerOptions: {
        default: {
          runner: '@nrwl/workspace/tasks-runners/default',
          options: {
            cacheableOperations: ['build', 'lint', 'test', 'e2e'],
          },
        },
      },
      projects: {},
    });
  });

  it('should create a prettierrc file', async () => {
    workspaceGenerator(tree, {
      name: 'proj',
      cli: 'nx',
      layout: 'apps-and-libs',
      defaultBase: 'main',
    });
    expect(tree.read('.prettierrc').toString()).toMatchSnapshot();
  });

  it('should recommend vscode extensions', async () => {
    workspaceGenerator(tree, {
      name: 'proj',
      cli: 'nx',
      layout: 'apps-and-libs',
      defaultBase: 'main',
    });
    const recommendations = readJson<{ recommendations: string[] }>(
      tree,
      '/.vscode/extensions.json'
    ).recommendations;

    expect(recommendations).toEqual([
      'ms-vscode.vscode-typescript-tslint-plugin',
      'esbenp.prettier-vscode',
    ]);
  });

  it('should recommend vscode extensions (angular)', async () => {
    workspaceGenerator(tree, {
      name: 'proj',
      cli: 'angular',
      layout: 'apps-and-libs',
      defaultBase: 'main',
    });
    const recommendations = readJson<{ recommendations: string[] }>(
      tree,
      '/.vscode/extensions.json'
    ).recommendations;

    expect(recommendations).toEqual([
      'nrwl.angular-console',
      'angular.ng-template',
      'ms-vscode.vscode-typescript-tslint-plugin',
      'esbenp.prettier-vscode',
    ]);
  });

  it('should add decorate-angular-cli when used with angular cli', async () => {
    workspaceGenerator(tree, {
      name: 'proj',
      cli: 'angular',
      layout: 'apps-and-libs',
      defaultBase: 'main',
    });
    expect(tree.exists('/decorate-angular-cli.js')).toBe(true);
    const packageJson = readJson(tree, '/package.json');
    expect(packageJson.scripts.postinstall).toEqual(
      'node ./decorate-angular-cli.js'
    );
  });

  it('should not add decorate-angular-cli when used with nx cli', async () => {
    workspaceGenerator(tree, {
      name: 'proj',
      cli: 'nx',
      layout: 'apps-and-libs',
      defaultBase: 'main',
    });
    expect(tree.exists('/decorate-angular-cli.js')).toBe(false);
    const packageJson = readJson(tree, '/package.json');
    expect(packageJson.scripts.postinstall).toBeUndefined();
  });

  it('should create a workspace using package layout', async () => {
    workspaceGenerator(tree, {
      name: 'proj',
      cli: 'nx',
      layout: 'packages',
      defaultBase: 'main',
    });
    expect(tree.exists('/packages/.gitkeep')).toBe(true);
    expect(tree.exists('/apps/.gitkeep')).toBe(false);
    expect(tree.exists('/libs/.gitkeep')).toBe(false);
    const nx = readJson(tree, '/nx.json');
    expect(nx.workspaceLayout).toEqual({
      appsDir: 'packages',
      libsDir: 'packages',
    });
  });
});
