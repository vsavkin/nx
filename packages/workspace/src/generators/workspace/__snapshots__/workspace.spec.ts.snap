// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nrwl/workspace:workspace should create a prettierrc file 1`] = `
"{
  \\"singleQuote\\": true
}"
`;
